torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2679,6 +2854,20 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None):
 
         expr = orig_expr
 
+        if hint is None:
+            concrete_val = self.size_hint(expr)
+        else:
+            concrete_val = sympy.sympify(hint)
+
+        if concrete_val is sympy.true:
+            g = expr
+        elif concrete_val is sympy.false:
+            g = sympy.Not(expr)
+        else:
+            g = sympy.Eq(expr, concrete_val)  # type: ignore[arg-type]
+
+        self._add_input_guard(g)
```

Comment:
Hmm, this may be fine for a start, but if you are just using the input expr as is, you may be too "late"; the guard will be in terms of an already simplified expression (that may have been simplified wrong). To truly track the original expressions, we need a variant that does NO simplification at all, not even sympy simplifications (which have been the source of some of our bugs!) In other words, we need to track the exact original Python operator calls that fed into the guard so we can record it faithfully.

Reply:
Hmmm... Could you give me some pointers? Where should I go look for them?

Reply:
There just... isn't any precedent for this. Maybe we can hop through a call and I can reason through it with you.

Reply:
Yukio's going to work in https://github.com/pytorch/pytorch/compare/main...ezyang:pytorch:shape-scratch?expand=1 to do this